src/concrete/ml/sklearn/tree_to_numpy.py

```diff
@@ -333,6 +345,121 @@ def tree_to_numpy(
     # but also rounding the threshold such that they are now integers
     q_y = tree_values_preprocessing(onnx_model, framework, output_n_bits)
 
-    _tree_inference, onnx_model = get_equivalent_numpy_forward_from_onnx(onnx_model)
+    _tree_inference, onnx_model = get_equivalent_numpy_forward_from_onnx(onnx_model, lsbs_to_remove)
 
     return (_tree_inference, [q_y.quantizer], onnx_model)
+
+
+# Remove this function once the truncate feature is released
+# FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4143
+def compute_lsb_to_remove_for_trees(onnx_model: onnx.ModelProto, q_x: numpy.ndarray) -> List[int]:
+    """Compute the LSB to remove for the comparison operators in the trees.
+
+    Referring to this paper: https://arxiv.org/pdf/2010.04804.pdf, there are
+    2 levels of comparison for trees, one at the level of X.A < B and a second at
+    the level of I.C == D.
+
+    Args:
+        onnx_model (onnx.ModelProto): The model to clean
+        q_x (numpy.ndarray): The quantized inputs
+
+    Returns:
+        List: the number of LSB to remove for level 1 and level 2
+    """
+
+    def get_bitwidth(array: numpy.ndarray) -> int:
+        """Compute the bitwidth required to represent the largest value in `array`.
+
+        Args:
+            array (umpy.ndarray): The array for which the bitwidth needs to be checked.
+
+        Returns:
+            int: The required bits to represent the array.
+        """
+
+        max_val = numpy.max(numpy.abs(array))
+        # + 1 is added to include the sign bit
+        bitwidth = math.ceil(math.log2(max_val + 1)) + 1
+        return bitwidth
+
+    def get_lsbs_to_remove(array: numpy.ndarray) -> int:
+        """Update the number of LSBs to remove based on overflow detection.
+
+        Args:
+            array (umpy.ndarray): The array for which the bitwidth needs to be checked.
+
+        Returns:
+            int: The updated LSB to remove.
+        """
+
+        initial_bitwidth = get_bitwidth(array)
+        # The subtraction operation increases precision by 1 or 2 bits
+        lsbs_to_remove = initial_bitwidth + 2
+
+        while lsbs_to_remove > 0:
+            half = 1 << (lsbs_to_remove - 1)
+            if get_bitwidth(array - half) <= initial_bitwidth:
+                lsbs_to_remove -= 1  # pragma: no cover
```

Comment:
I think something is fishy here : we have an if-else statement bu both outcomes have `# pragma: no cover` ? both should be tested !